### ISSUE TITLE
Support a wider range of cucumber-gherkin versions

### ISSUE THREAD

**Issue by enkessler** [cucumber/cucumber-ruby#1453]
**Is your feature request related to a problem? Please describe.**
I have a project that has, as a direct dependency, the `cucumber-gherkin` gem. I use the `cucumber` gem, which also has a direct dependency on `cucumber-gherkin`, to test my project. The problem is that, whereas my project supports a wide range of versions of `cucumber-gherkin` (and I therefore must test against each of them), `cucumber` only supports a single major version of `cucumber-gherkin` at a time and there are even some major versions of `cucumber-gherkin` that it skips over entirely.

The result of this is that I have to constantly switch which version of `cucumber` I am using when testing and, more often than not, use extremely old versions of `cucumber` (e.g. version 2.2.0) because it is *not possible* to use any recent version of `cucumber` to test my project for any version of `cucumber-gherkin` that does not have a corresponding `cucumber` version that supports it. 

**Describe the solution you'd like**
Release at least one version of the `cucumber` gem for each major version of the `cucumber-gherkin` gem.

**Describe alternatives you've considered**
As far as I understand the nature of gem versions and maintaining a sane gem ecosystem, there is no alternative solution. Using the `cucumber` to test my project will lock me into exactly one major version of `cucumber-gherkin` and if the version of `cucumber-gherkin` that I need to load into memory for testing is not that version then I am out of luck.

**Comment by enkessler** [comment]
I'd rest a lot easier knowing that Cucumber will still be a viable testing tool for me in the future.

**Comment by enkessler** [comment]
`cucumber-gherkin` 16 and 17 have come out and the most recent `cucumber` is still only compatible with 15.

**Comment by luke-hill** [comment]
Most of the time the cucumber ruby gem will only work with 1 (major), version of gherkin. Because the two are quite tightly coupled.

For your framework/s you should be able to get around this quite easily by allowing a variety of versions of the cucumber gem. And dependent on each version of cucumber you support, you then get different allowances.

If that's proving too tricky, you could flip it, and instead just support a variety of versions of gherkin, and then let that backwards fill the support for the versions of cucumber. I don't really see this being an issue though?

**Comment by enkessler** [comment]
> Most of the time the cucumber ruby gem will only work with 1 (major), version of gherkin.

Right. That's the problem.

> Because the two are quite tightly coupled.

Why? Tight coupling like that is generally considered A Bad Thing.

> For your framework/s you should be able to get around this quite easily by allowing a variety of versions of the cucumber gem. And dependent on each version of cucumber you support, you then get different allowances.

> If that's proving too tricky, you could flip it, and instead just support a variety of versions of gherkin, and then let that backwards fill the support for the versions of cucumber. I don't really see this being an issue though?

Quite easily? Not really an issue? @luke-hill I don't think that you understand the situation that I am describing. This is not a problem with a "tricky" solution. This is a technical impossibility imposed by Cucumber.

**Comment by enkessler** [comment]
> But gherkin is designed to be consumed by the cucumber library in the versions specified. In the same way a private method or function in a class can be thought of to be only consumable by it's public callers.

Are you suggesting that `cucumber-gherkin` should not actually be treated like any other public facing gem? That it should be used by no one but `cucumber`?

**Comment by luke-hill** [comment]
I'm suggesting that is what it is primarily designed for. If you wish to use it for other things that is fine. But it is open source. So people are able to view / edit the source code and/or fork the gem and make suggestions for modifications.

For example, if you edit / amend the gemspec to allow the gem to work on other versions of cucumber, maybe some bits we have designed won't work. But then maybe those bits that won't work aren't the bits you care about in your package (I don't know).

This is a discussion we're having RE protobuf at the moment (As a real world example), as this loads in one of the rails auto-loaded libraries. However we only use a small proportion so maybe we can amend/remove it (I don't know I've not dug into it). This would be handy as it would remove a heavy chunk of code from a package we only partially use.

Maybe for your library you could do something similar? Like we have done. See https://rubygems.org/gems/protobuf-cucumber for what the team have done

**Comment by enkessler** [comment]
The beauty of open source if that the user gets to do all of the work themselves...yeah, no.

If I, as a user, have to jump through a bunch of advanced programming hoops in order to use Cucumber, then Cucumber has already lost a significant portion of its value as a useful tool.

I can make my library work with every version of `gherkin`/`cucumber-gherkin` from 2 to 16. Been doing it for years. It's not difficult, just occasionally tedious. What I want is for `cucumber` to do the same. Heck, if the tight coupling is intended then I would _expect_ a new `cucumber` release to go along with every `cucumber-gherkin` release. That would be 1:1. Currently, the ratio is 3:5.

**Comment by enkessler** [comment]
To give you a better idea of what my situation is, here is what had to be done to make `cuke_modeler` compatible with `cucumber-gherkin` 16: https://github.com/enkessler/cuke_modeler/commit/9dea9b87158a5e7b482fbb9773e80b904cea9e86

Pretty straightforward, yes? I just updated some spots in the code that are version specific to be aware of the new version and make a new adapter for that major version. Note that the 16.x adapter is really just the 9.x adapter. **Spoiler**: _all_ of the current adapters are just the 9.x adapter in disguise. That's how little has changed in the past eight major release of `cucumber-gherkin` from the point of view of my gem.

But also note that I am using `cucumber 2.2.0` for most of my testing. I have to do that because it is _Cucumber_ that is the one not supporting enough versions of Gherkin.

**Comment by aurelien-reeves** [comment]
There may be something to improve here indeed

I took a look in the changelog of gherkin. From version 9, there were two reasons for new major versions:
- updated of the `messages` dependency
- breaking changes in the Javascript implementation

I am not sure how `messages` may impact `gherkin` from the outside world perspective.
But definitely, changes in the Javascript implementation resulting in a new major release of the ruby gem may seems not that obvious.

If we should consider all those depencencies and impact for each release of each module, we may spent most of the time just doing this.

Is there something we should / could do to improve all those dependencies without losing our mind in the process? Could we consider not releasing version for a flavour which has actually no changes in it or would that be too difficult to manage? Anyone would have suggestion? I will document myself on the subject in the meantime.

Cc. @luke-hill @aslakhellesoy

**Comment by luke-hill** [comment]
I think the issue here is (as mentioned earlier), that how does one think of the various cucumber sub-gems (In ruby, and potentially other languages).

So for example cucumber-core and cucumber-wire are designed to be extracted pieces of code that are ran in tandem with cucumber. I think having multiple versions of gherkin supported in the cucumber implementation is possible to an extent, but it would dramatically increase the workload.

Furthermore, there are downstream consumers of our tool (Such as RubyMine / IJ), which rely on us using a specific version of cucumber / gherkin at a specific time. Once we start adding more flexibility, their rigid processes will become more clunky. I'm not saying that has to be our highest priority/consideration, but it is _**a**_ consideration.

Also some major versions of gherkin are inherently undesirable because they were short lived and not tested fully, because I know we did a phase where we bumped 2 or 3 majors quite quickly. For instance we went from 9.2 -> 13 in under 3 months. So whilst the quantity of API breaking changes may suggest 3 major version differences, the changelog and diffs may suggest a small quantity of changes.

Using the gherkin-ruby read only I got this (For 4 major version changes)

```
76 changed files with 892 additions and 388 deletions.
```

That's not a super huge change for 4 majors IMO. And in that time there wasn't a single stable cucumber release consuming them. Only 1 rc version of cucumber consuming them (I know from first hand exp, that some of the cucumber rc's weren't fully working).

**Comment by enkessler** [comment]
> I think the issue here is (as mentioned earlier), that how does one think of the various cucumber sub-gems (In ruby, and potentially other languages).
>
> So for example cucumber-core and cucumber-wire are designed to be extracted pieces of code that are ran in tandem with cucumber

I think of them as independent tools. Perhaps they are tools with a very specific purpose which possibly limits how many consumers will exist that need that particular functionality (which is why half of the Cucumber tech stack is only used when the consumer is also using Cucumber) but they are developed, maintained, and distributed independently, all the same.

> I think having multiple versions of gherkin supported in the cucumber implementation is possible to an extent, but it would dramatically increase the workload.

To put a bit of a fine point on it: So what? Don't be lazy.

The nature of dependencies is such that any dependency that a gem has is necessarily passed on to the consumers of that gem. Heck, we have an entire tool (Bundler) that exists solely to deal with the massive and intertwined dependency tree that can spring into existence just by explicitly including one single gem in a project. This is why the general rule of having dependencies is to have

1. As few of them as possible and
2. be compatible with as broad a range of versions of those dependencies as possible

Could I use the `backports` gem or `activesupport` gem in order to have more convenient code for my gem? Yes. Am I going to use those gems in my gem? No. Because as soon as I do then every project that includes my gem must also include those gems and that could cause problems. Instead I go through the effort of writing Ruby code that is still valid on older interpreters and I write little helper methods that accomplish the goal of convenience without having to monkey patch (read: pollute) the execution environment. And for the dependencies that I do have, I make sure that I am compatible with several different major versions of them because I have no idea what the environmental limitations of my consumers are going to be.

> Furthermore, there are downstream consumers of our tool (Such as RubyMine / IJ), which rely on us using a specific version of cucumber / gherkin at a specific time. Once we start adding more flexibility, their rigid processes will become more clunky. I'm not saying that has to be our highest priority/consideration, but it is a consideration.

Hi! Downstream consumer of your tool, here. *waves* I can tell you for a fact that my process is clunky because of a _lack_ of flexibility, not because you've added too much flexibility. Please consider me.

RubyMine is great and the IntelliJ folks make the best tools and I use them all the time. That being said, if the failings of their tools are resulting in them dictating terms to Cucumber instead of the other way around then something has gone horribly wrong.

> Also some major versions of gherkin are inherently undesirable because they were short lived and not tested fully, because I know we did a phase where we bumped 2 or 3 majors quite quickly. For instance we went from 9.2 -> 13 in under 3 months. So whilst the quantity of API breaking changes may suggest 3 major version differences, the changelog and diffs may suggest a small quantity of changes.

Setting aside the facepalming situation that is 'people who make a testing tool not bothering to test properly', those versions exist. They are available for download. They are used and relied upon in production environments. If something is wrong with them then make a new patch release that fixes the bugs. Or yank the version and stop providing it for consumption. Maybe update the CHANGELOG with a big old message saying "THIS VERSION IS BAD AND SHOULD NOT BE USED". Don't just quietly not support it and hope that people never use it.

> That's not a super huge change for 4 majors IMO.

`<sarcasm>`Then it should be relatively easy for Cucumber to support all 4 of those major versions, shouldn't it?`</sarcasm>`

> And in that time there wasn't a single stable cucumber release consuming them

Yes. This is the crux of the issue. Please have the `cucumber` gem support more versions of the `cucumber-gherkin` gem.

**Comment by enkessler** [comment]
`cucumber-gherkin` 18 has just released. The usability to un-usability ratio for `cucumber` has now dropped from 3:5 to 3:6.

**Comment by aslakhellesoy** [comment]
Why don’t you send a pull request to address your concerns @enkessler?

I also suggest you to re-read the license.

**Comment by enkessler** [comment]
> Why don’t you send a pull request to address your concerns @enkessler?

For the same reason that I haven't fixed most of the other issues that I've found: I can't get the project to build on my machine. The switch to a Docker environment was an improvement from me needing a separate machine for Cucumber development but it still doesn't build cleanly when I try it.

> I also suggest you to re-read the license.

I'm aware that this is a free product maintained by a bunch of volunteers and provided without any guarantee of anything at all, yes. But...*waves his hand vaguely at the mountain of books and articles published about Cucumber, the educational courses and paid consultations, the wide industry adoption, the corporate acquisition/sponsorship, etc.* ...surely, @aslakhellesoy , it is at least understandable why I try and hold this project to a higher standard than the obligation that is officially declared?

**Comment by aslakhellesoy** [comment]
> I can't get the project to build on my machine.

Then I suggest you ask for help about that. Either in Slack, or open a GitHub issue.

> it is at least understandable why I try and hold this project to a higher standard than the obligation that is officially declared?

Absolutely not. 

The majority of the contributors are still volunteers. The exponential growth of Cucumber adoption has not led to an exponential growth of contributors. This means there is more work than we can handle.

The volunteer contributors mostly scratch their own itch. They owe you nothing @enkessler. Your problem is your problem, not theirs.

As for me and the two other people who are paid to work on Cucumber, we prioritise issues that we believe affect a large number of users. I'm not convinced this is one of them. Your attitude does not help in giving this priority, quite the opposite.

And even if we were an army of developers working to fix all bugs, you don't get to complain if we don't fix yours. Not here anyway.

**Comment by mattwynne** [comment]
@enkessler what's the dependency you have on the `cucumber` gem for? Is it only for running tests against your project? (i.e. a dev dependency) or do you also use parts of the `cucumber` gem in the runtime code in your project too?

**Comment by enkessler** [comment]
@aslakhellesoy 
> Then I suggest you ask for help about that. Either in Slack, or open a GitHub issue.

I did ask for help. https://cucumberbdd.slack.com/archives/C62GZFLLT/p1597690552004000
To your credit, you responded. Unfortunately, things didn't go much further and no one else chimed in. Because this isn't the first time that I've run into issues trying to hack on the project myself and because people probably have better uses for their time than trying to get one problematic contributor up and running, I tend to just let the local development issue drop after a while.

>> it is at least understandable why I try and hold this project to a higher standard than the obligation that is officially declared?

>Absolutely not.

Oh. Okay. That's not the stance that I was expecting. Maybe we aren't on the same page. It wouldn't be the first time that I've been in a conversation with someone and we each experienced an entirely different conversation than the other despite each thinking that our version to lined up fine with the words being said.

> The majority of the contributors are still volunteers. The exponential growth of Cucumber adoption has not led to an exponential growth of contributors. This means there is more work than we can handle.

Aye, and they seem to be a good sort of lot. Quality and not quantity, and I understand that that means prioritizing the work that does exist. 

> The volunteer contributors mostly scratch their own itch. They owe you nothing @enkessler.

I'd happily scratch this one myself if I could. I know that no one owes me anything. If this project closed down tomorrow then all that I could do is shrug and say "Well, that was an interesting chapter of my life.". 

>  Your problem is your problem, not theirs.

As is the case for anyone's problems. No user has to bring up issues and no maintainer has to deal with them. And yet the issue tracking systems, the forums, and the community at large continue to exist. I'm not here to badger people into fixing stuff. I'm here to add additional (and sometimes poorly communicated) input into the collective.

> As for me and the two other people who are paid to work on Cucumber, we prioritise issues that we believe affect a large number of users. I'm not convinced this is one of them.

Definitely. This is an edge case. It is a _crippling_ edge case but it is definitely outside of the norm.

> Your attitude does not help in giving this priority, quite the opposite.

I am fully aware that I make a horrible poster child for a cause. Doubly so on the internet.

> And even if we were an army of developers working to fix all bugs, you don't get to complain if we don't fix yours. Not here anyway.

I'm not complaining that people aren't fixing the bugs that I find. I'm still at the stage of trying to clarify what problem even is and ensure that people understand the nature of it and its wider ramifications and, based on some of the responses that I've gotten, that hasn't seemed to be the case. If, after that, it is decided that the problem is not important enough to deal with then that is fine. Close the issue or slap a 'backlog' label on it or whatever. "Finite resources" and all that. It happens. I'll understand.

As far as my tone...yeah, that's a fair callout. Honestly, I expected you to throw the Code of Conduct in my face instead of the License Agreement. If you'll pardon the paraphrasing, when I interpret an answer as "that would take effort and we have other places to spend that effort" then that's perfectly fine. It is when an the answer comes across as "that would take effort and effort is Not A Thing That We Do" that ire is raised. Presumably others are similarly affected when someone (read: me) comes across as "whiny, entitled git" instead of "well-intentioned assistant".

**Comment by enkessler** [comment]
@mattwynne 

> @enkessler what's the dependency you have on the `cucumber` gem for? Is it only for running tests against your project? (i.e. a dev dependency) or do you also use parts of the `cucumber` gem in the runtime code in your project too?

The `cucumber` gem is only a development dependency for me. I _could_, in theory, just forgo using Cucumber altogether and turn my feature files from "executable documentation" into the regular "just documentation" kind. The `cucumber-gherkin` gem is a runtime dependency for me. Being an abstraction layer on top of that is _the entire purpose_ of my gem. That's why I'd drop Cucumber and just test with only RSpec before I would ever not test against one of the major versions of Gherkin.

**Comment by enkessler** [comment]
Tangential conversation:

Regardless of whether or not the `cucumber` gem supports more than one major version of `cucumber-gherkin` at a time, not changing the major version of `cucumber` when it's dependency limitations change is (debatably) [breaking semantic versioning](https://github.com/semver/semver/issues/148).

**Comment by luke-hill** [comment]
So long as the cucumber API isn't changing then you're fine to release things as a minor x.Y.z

The next big release that will happen to cucumber is the introduction of rule tagging (Which is definitely backwards compatible
in terms of the cucumber API)

**Comment by mattwynne** [comment]
@enkessler

> The `cucumber` gem is only a development dependency for me.

OK. So I have an experience here that might be helpful.

Back when we re-wrote the internals of cucumber-ruby as the cucumber-ruby-core gem, we still needed to be able to run Cucumber's own regression suite of Gherkin acceptance tests against the re-write. Initially, it was impossible to use the new code to actually run the tests, as it wasn't nearly reliable enough yet. A chicken-and-egg problem.

So we worked out a way to use the current stable version of Cucumber to execute the tests, but have them run their tests on the development version of Cucumber.

I can't remember exactly how we did it, but I think it involved us using separate Ruby processes for the test runner (using the stable Cucumber) and the code invoked by the tests.

Maybe you could do something like that too? I honestly think that's a more likely route to solving your problem than trying to make Cucumber itself backwards compatible with all these old versions of Gherkin.

**Comment by enkessler** [comment]
@luke-hill Yep, that's essentially the 'letter of the law' side of the debate. Me, I'm more of a 'spirit of the law' kind of person. If a tool goes from being perfectly usable to completely unusable then that's about as backwards incompatible as it can get, even if a strict reading of the definition of Semantic Versioning does say that it's fine for such a change to only be a patch level release because method signatures weren't involved.

Anyway, that was more of a side note about the paperwork involved in dropping support for a dependency. The main topic at hand remains on whether or not to have that support. So...

@mattwynne If it were just me then maybe such a solution would work. Mind you, _I_ still wouldn't do it because that's more effort than it's worth and I would rather just keep using `cucumber` `2.2.0` and then simply stop using Cucumber altogether once that version becomes unusable (either because the gem ecosystem moves on so much that other dependency interactions start being problematic or because I want to use newer features of the Gherkin language like the `Rule` keyword). But it's _not_ (in theory) just me. This is a problem that would affect every project with `cucumber-gherkin` as a direct runtime dependency. "Peer dependency" I think is the term used (mostly in NPM land). Having each and every downstream consumer implement a work-around makes less sense than solving the problem at its source.

Hmm. Although I suppose that any consumer of `cucumber-gherkin` could, in turn, become the 'source' of the problem for any and all other consumers by introducing extremely restrictive dependency version limits. (*wanders over to RubyGems*) Well, what a small world we few [consumers ](https://rubygems.org/gems/cucumber-gherkin/reverse_dependencies) are. `cucumber` just happens to be the most widely used consumer and so, yeah, it's no surprise that it's the most likely candidate for encountering a version limit conflict. Because I use both RSpec and Cucumber to test my gem, if I decided to slap `turnip` on top of `rspec`, I'd quickly run into this same fundamental problem (only worse) and, yes, I'd be just as annoyed at the dependency limit conflicts due to `turnip` as I am when I encounter them due to `cucumber`. As it stands, `cuke_modeler` does appear to be the only downstream gem with an eccentric enough maintainer to not lock down its peer dependencies so tightly... Fine, I admit it: I want the world to be a better place and enjoy making my tools play nicely with others.

But, hey, the future is unknown. Maybe other projects will come along and join the club. Maybe there are projects out there that aren't published gems but do use `cucumber-gherkin` directly and, in doing so, they are super secret club members that the rest of us don't know about. It also seems that avoiding the problem can only be solved by the consumer projects themselves. So the questions is: What is the best way for those consumers to solve the problem? Is it for each consumer to use esoteric programming techniques to somehow isolate every single dependency so that none of them can ever be a problem for themselves? Is it for each consumer to try and be flexible enough with their dependencies that they are never a problem for others? Is there, looking at the problem from this new, broader perspective of software library development in general, some other solution?

Personally, between the figurative choices of "I'm going to build an impenetrable survival bunker so that nothing can hurt me" and "I'm going try and build a world where survival bunkers aren't needed", I'd choose the latter. But I'm weird like that.

**Comment by mattwynne** [comment]
OK, I would like to draw this discussion to a close. As I see it, there are four options:

1) cuke-modeller stops supporting legacy gherkin versions.
2) cuke-modeller stops using Cucumber to test itself and uses a different test runner that doesn't depend on gherkin.
3) cuke-modeller uses a separate Ruby process to execute Cucumber tests, or separate Gemfiles for testing legacy gherkin versions.
4) cucumber-ruby is adapted to work with multiple legacy versions of gherkin.

While I think it's laudable for you to be supporting legacy gherkin versions with your tool, do you know how many of your users rely on this capability? Perhaps it's worth considering whether the benefits outweigh the cost.

Certainly, from a Cucumber point of view, I would not like to see us support legacy versions of Gherkin. There's a reason we've left those old versions behind, and introducing the extra adapter code to keep things running with the different versions would be a maintenance nightmare. I realise it's a hassle for tool-builders like yourself when the API changes so much as it has, and I'm optimistic that, with the current architecture using the message protocol and Berp parser, we've started to stabilise on something that we can now refine rather than reinvent.

**Comment by enkessler** [comment]
> OK, I would like to draw this discussion to a close.

Yes. This horse is well and truly dead. I'll give it my one last kick and then be on my way.

> As I see it, there are four options:
> 
> 1. cuke-modeller stops supporting legacy gherkin versions.
> 2. cuke-modeller stops using Cucumber to test itself and uses a different test runner that doesn't depend on gherkin.
> 3. cuke-modeller uses a separate Ruby process to execute Cucumber tests, or separate Gemfiles for testing legacy gherkin versions.
> 4. cucumber-ruby is adapted to work with multiple legacy versions of gherkin.

1. Nah. The support is already there. Removing it would actually take more work than to continue supporting it. Besides, the difficulty doesn't come from the release's age. It comes from whether or not other tools ever supported it. From `cuke_modeler`'s point of view, the current version `18.0` of `cucumber-gherkin` is equally as 'difficult' to support as the legacy version `11.0` of `cucumber-gherkin` because both of those versions are not supported by `cucumber`.
2. That's an option. It's a worse case scenario that I'd rather avoid because, you know, _Cucumber is a nice tool to have_ (congratulations on the culmination of your years of effort, by the way). I'd not be fighting for it if it weren't. Realistic worse case: I'll adjust my test suites so that only the RSpec stuff runs on unsupported versions and both RSpec and Cucumber run on the supported versions. The resulting increased (but still negligible) testing gap is an acceptable trade off in order to continue use of Cucumber in a limited capacity.
3. I already use separate Gemfiles in order to use different versions of `cucumber` as needed. That's how I've gotten away with things for this long (that, and never using the `Rule` keyword). Regarding using a separate Ruby process for testing with Cucumber...that's doable. As far as building survival bunkers go, the memory barrier of a separate process is a lot more reliable of a route than whatever loss of fidelity that meta-programming shenanigans would have resulted in. That being said, reworking my step definitions to do external process stuff remains less appealing that just not running the Cucumber tests against unsupported versions. It's at times like this that I'm glad that I made the decision to ensure 100% coverage from the RSpec suite alone in order to have the freedom to alter my Cucumber documentation without fear of loosing functional coverage. I just wasn't ever ever expecting the 'alteration' in question to be 'not running them'. Hooray for silver linings!
4. This is the least likely but most ideal situation, yes. 

> While I think it's laudable for you to be supporting legacy gherkin versions with your tool, do you know how many of your users rely on this capability?

Honestly, I don't even know that my tools have much in the way of users beyond myself and whatever work projects I happen to get involved with. `cuke_modeler` and its miniature ecosystem of gems have no community that I am aware of. I can see download counts and reverse dependencies but I couldn't tell you if I'm providing a valuable service to the wider world or just making myself feel good by enabling some hypothetical programmer somewhere. I can tell you with some confidence, however, that users are strange creatures and expecting them all to keep up with the times and always upgrade as things become available is not realistic. That's why Long Term Support releases are a thing and why it is generally a good idea to be flexible when declaring version limitations.

`<friendly ribbing>` Speaking of laudably supporting legacy versions of dependencies, it's good to see that `cucumber` is still supporting version 15.x of `cucumber-gherkin`. That version is over 6 months old, which is practically ancient by `cucumber-gherkin` standards. ;) `</friendly ribbing>`

> Perhaps it's worth considering whether the benefits outweigh the cost.

The benefit is that I continue to learn new tricks and techniques and become a better developer. As I mentioned before, it's now quite easy to update `cuke_modeler` when a new version of `cucumber-gherkin` is released. It took some thought and experimentation to get to that point but now that I'm there, the effort has paid off.

The cost is that I had to spend time learning. I'll happily make that trade.

> Certainly, from a Cucumber point of view, I would not like to see us support legacy versions of Gherkin. There's a reason we've left those old versions behind, and introducing the extra adapter code to keep things running with the different versions would be a maintenance nightmare.

**Reminder**: I'm _not_ asking that `cucumber` support all of those versions _at the same time_. As long as there is even a single release of `cucumber` that supports any given major version of `cucumber-gherkin` then the problem is solved. `cucumber` is free to drop support for legacy versions of dependencies the moment that they become so. No adapters or any out of the ordinary upgrading effort is required. Just don't avoid upgrading for so long that the current versions of your dependencies (e.g. `cucumber-gherkin`) _become_ the legacy versions by the time that you do upgrade.

> I realise it's a hassle for tool-builders like yourself when the API changes so much as it has, and I'm optimistic that, with the current architecture using the message protocol and Berp parser, we've started to stabilise on something that we can now refine rather than reinvent.

Oh, I solved that problem the day that Gherkin 3.x was released. `cucumber-gherkin` can release a new backwards incompatible API every month for all I care and my gems will adapt. I've solved that problem. It's those _other_ tool-builders that I'm worried about and have been trying to give a pep talk to. ;)

But, yeah, if the release rate of `cucumber-gherkin` slows down then the needed dependency update frequency for `cucumber` is likely to drop low enough to become manageable for the maintainers. Then we can all become compatible gem friends again.


Cheers!
